File: python/utils/units.py
from __future__ import annotations

import re

from errors import BridgeError

EMU_PER_INCH = 914400
EMU_PER_POINT = 12700
EMU_PER_CM = 360000
EMU_PER_PIXEL = 9525

_MEASUREMENT_RE = re.compile(r"^(\d+(?:\.\d+)?)(in|pt|cm|px)$", re.IGNORECASE)


def to_emu(value: int | float | str) -> int:
    if isinstance(value, int | float):
        return int(round(value))

    if not isinstance(value, str):
        raise BridgeError(
            code="validation_error",
            message="Measurement must be number (EMU) or string with unit.",
            details={"value": value},
        )

    match = _MEASUREMENT_RE.match(value.strip())
    if not match:
        raise BridgeError(
            code="validation_error",
            message="Invalid measurement string.",
            details={"value": value, "expected": "2in, 24pt, 5cm, 96px"},
        )

    numeric = float(match.group(1))
    unit = match.group(2).lower()

    if unit == "in":
        return int(round(numeric * EMU_PER_INCH))
    if unit == "pt":
        return int(round(numeric * EMU_PER_POINT))
    if unit == "cm":
        return int(round(numeric * EMU_PER_CM))
    if unit == "px":
        return int(round(numeric * EMU_PER_PIXEL))

    raise BridgeError(
        code="validation_error",
        message="Unsupported measurement unit.",
        details={"unit": unit},
    )


def emu_to_inches(value: int | float) -> float:
    return float(value) / EMU_PER_INCH

File: src/bridge/units.ts
const EMU_PER_INCH = 914400;
const EMU_PER_POINT = 12700;
const EMU_PER_CM = 360000;
const EMU_PER_PIXEL = 9525;

export type MeasurementInput = number | string;

const MEASUREMENT_RE = /^(\d+(?:\.\d+)?)(in|pt|cm|px)$/i;

export function measurementToEmu(value: MeasurementInput): number {
  if (typeof value === "number" && Number.isFinite(value)) {
    return Math.round(value);
  }

  if (typeof value !== "string") {
    throw new Error("Measurement must be a number (EMU) or string with unit");
  }

  const input = value.trim().toLowerCase();
  const match = input.match(MEASUREMENT_RE);
  if (!match) {
    throw new Error(`Invalid measurement '${value}'. Expected format like 2in, 24pt, 5cm, or 96px.`);
  }

  const numeric = Number(match[1]);
  const unit = match[2];

  switch (unit) {
    case "in":
      return Math.round(numeric * EMU_PER_INCH);
    case "pt":
      return Math.round(numeric * EMU_PER_POINT);
    case "cm":
      return Math.round(numeric * EMU_PER_CM);
    case "px":
      return Math.round(numeric * EMU_PER_PIXEL);
    default:
      throw new Error(`Unsupported measurement unit '${unit}'.`);
  }
}

export function emuToInches(emu: number): number {
  return emu / EMU_PER_INCH;
}
